pkg/cluster/operation/action.go

```diff
@@ -191,77 +223,80 @@ func Restart(
 }
 
 // StartMonitored start BlackboxExporter and NodeExporter
-func StartMonitored(ctx context.Context, instance spec.Instance, options *spec.MonitoredOptions, timeout uint64) error {
-	ports := map[string]int{
-		spec.ComponentNodeExporter:     options.NodeExporterPort,
-		spec.ComponentBlackboxExporter: options.BlackboxExporterPort,
+func StartMonitored(ctx context.Context, hosts []string, options *spec.MonitoredOptions, timeout uint64) error {
+	return systemctlMonitor(ctx, hosts, options, "start", timeout)
+}
+
+// StopMonitored stop BlackboxExporter and NodeExporter
+func StopMonitored(ctx context.Context, hosts []string, options *spec.MonitoredOptions, timeout uint64) error {
+	return systemctlMonitor(ctx, hosts, options, "stop", timeout)
+}
+
+// EnableMonitored enable/disable monitor service in a cluster
+func EnableMonitored(ctx context.Context, hosts []string, options *spec.MonitoredOptions, timeout uint64, isEnable bool) error {
+	action := "disable"
+	if isEnable {
+		action = "enable"
 	}
-	e := ctxt.GetInner(ctx).Get(instance.GetHost())
+
+	return systemctlMonitor(ctx, hosts, options, action, timeout)
+}
+
+func systemctlMonitor(ctx context.Context, hosts []string, options *spec.MonitoredOptions, action string, timeout uint64) error {
+	ports := monitorPortMap(options)
 	for _, comp := range []string{spec.ComponentNodeExporter, spec.ComponentBlackboxExporter} {
-		log.Infof("Starting component %s", comp)
-		log.Infof("\tStarting instance %s", instance.GetHost())
-		c := module.SystemdModuleConfig{
-			Unit:         fmt.Sprintf("%s-%d.service", comp, ports[comp]),
-			ReloadDaemon: true,
-			Action:       "start",
-			Timeout:      time.Second * time.Duration(timeout),
-		}
-		systemd := module.NewSystemdModule(c)
-		stdout, stderr, err := systemd.Execute(ctx, e)
+		log.Infof("%s component %s", actionPrevMsgs[action], comp)
 
-		if len(stdout) > 0 {
-			fmt.Println(string(stdout))
-		}
-		if len(stderr) > 0 {
-			log.Errorf(string(stderr))
-		}
+		errg, _ := errgroup.WithContext(ctx)
+		for _, host := range hosts {
+			host := host
+			nctx := checkpoint.NewContext(ctx)
+			errg.Go(func() error {
+				log.Infof("\t%s instance %s", actionPrevMsgs[action], host)
+				e := ctxt.GetInner(nctx).Get(host)
+				service := fmt.Sprintf("%s-%d.service", comp, ports[comp])
 
-		if err != nil {
-			return errors.Annotatef(err, "failed to start: %s", instance.GetHost())
-		}
+				if err := systemctl(nctx, e, service, action, timeout); err != nil {
+					return toFailedActionError(err, action, host, service, "")
+				}
 
-		// Check ready.
-		if err := spec.PortStarted(ctx, e, ports[comp], timeout); err != nil {
-			return toFailedActionError(err, "start", instance)
-		}
+				var err error
+				switch action {
+				case "start":
+					err = spec.PortStarted(ctx, e, ports[comp], timeout)
```

Comment:
Use nctx otherwise the test will fail.

See https://github.com/pingcap/tiup/blob/master/doc/design/checkpoint.md#parallel-task

Reply:
Thanks, PTAL